### PR TITLE
Better logging message on header count exceeded

### PR DESCRIPTION
**Pull request by plallin**
Signed-off-by: Pauline <pauline.lallinec@gmail.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)

Commit Message: Better error message on header count exceeded
Additional Description: The header size and header count limit exceeded is the same message "header size exceeded". This PR clarifies the case when header count is exceeded. 
Risk Level: low
Testing: unit tests updated
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: source/common/http/http1/codec_impl.cc [hunk 1]
```diff
@@ -512,7 +512,7 @@ Status ConnectionImpl::completeLastHeader() {
     RETURN_IF_ERROR(sendProtocolError(Http1ResponseCodeDetails::get().TooManyHeaders));
     const absl::string_view header_type =
         processing_trailers_ ? Http1HeaderTypes::get().Trailers : Http1HeaderTypes::get().Headers;
-    return codecProtocolError(absl::StrCat(header_type, " size exceeds limit"));
+    return codecProtocolError(absl::StrCat(header_type, " count exceeds limit"));
```
Comment (asraa) [inline source/common/http/http1/codec_impl.cc:515]: does this suffice for your use case? if you want the difference to show up in the response (as opposed to details in logging), `error_code_` will be the response body. right now it is  `Http::Code::RequestHeaderFieldsTooLarge` for both.

**Review by asraa — COMMENTED** [review]
Change looks good to me! Let me know about that note, otherwise LGTM

File: source/common/http/http1/codec_impl.cc (hunk 1 above)
Comment (plallin) [reply source/common/http/http1/codec_impl.cc:515]: That's an excellent remark. My use case is not on requests, though, it's for responses. Where we had a response from a server sending a whooping 116 headers, the error in Envoy showed up as (1) headers disconnect before response and (2) 503 Upstream Disconnect. 

Only by increasing logging we could figure out that there were too many headers in the response, causing Envoy to terminate the connection.

Comment (asraa) [reply source/common/http/http1/codec_impl.cc:515]: Ah! Thanks for the clarification.

**Comment by asraa** [comment]
Could you please fix format? Otherwise LGTM
/wait

**Comment by plallin** [comment]
@asraa ty; I have failing tests I need to fix; I've not been able to execute them locally yet and the test runner provided doesn't give much details as to why they're failing. I'll need to get back to it this week-end...

**Comment by plallin** [comment]
@asraa I ?think? the tests passed. Can you help me check whether the failures in the pipeline are related?

I see 2 of them

```
2021-02-04T00:14:31.2653668Z spawn docker run --rm -ti --network=host solsson/websocat ws://localhost:10000
2021-02-04T00:14:31.5161823Z websocat: WebSocketError: I/O failure
2021-02-04T00:14:31.5166642Z websocat: error running
```

and

```
2021-02-03T23:23:18.4723173Z FAILED: //test/extensions/filters/http/grpc_json_transcoder:grpc_json_transcoder_integration_test (Summary)
2021-02-03T23:23:18.4726022Z       C:/build/tmp/execroot/envoy/bazel-out/x64_windows-opt/testlogs/test/extensions/filters/http/grpc_json_transcoder/grpc_json_transcoder_integration_test/test.log
2021-02-03T23:23:18.4732019Z       C:/build/tmp/execroot/envoy/bazel-out/x64_windows-opt/testlogs/test/extensions/filters/http/grpc_json_transcoder/grpc_json_transcoder_integration_test/test_attempts/attempt_1.log

```

Meanwhile the test I updated passed:

```
2021-02-03T23:28:57.3100752Z //test/common/http/http1:codec_impl_test                                 PASSED in 8.4s
```

and the `grpc_json_transcoder` integration test passed in other (non-windows) pipeline:

```
2021-02-03T23:14:49.9377086Z //test/extensions/filters/http/grpc_json_transcoder:grpc_json_transcoder_integration_test PASSED in 12.8s

```

(from [this pipeline](https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/65642/logs/56)

If this is unrelated to the current PR, is it possible to re-run the test?

If it is related, can you walk me through it? This is my first non-documentation PR in Envoy, please excuse the newbieness :-)

**Comment by plallin** [comment]
Looking at the failed script, it has had a lot of failures recently with only 1 out of 5 pipeline being successful.

I can see on a successful pipeline, it does:

```
021-02-02T13:00:09.4506146Z > [tls] Interact with web socket ws -> ws
2021-02-02T13:00:09.6171700Z spawn docker run --rm -ti --network=host solsson/websocat ws://localhost:10000
2021-02-02T13:00:10.6232867Z HELO
2021-02-02T13:00:10.6234284Z [ws] HELO
2021-02-02T13:00:10.6261953Z GOODBYE
2021-02-02T13:00:10.6262718Z [ws] HELO
2021-02-02T13:00:10.6270336Z 
2021-02-02T13:00:10.6272389Z > [tls] Interact with web socket wss -> wss
2021-02-02T13:00:10.6326487Z spawn docker run --rm -ti --network=host solsson/websocat --insecure wss://localhost:20000
2021-02-02T13:00:11.6409889Z HELO
2021-02-02T13:00:11.6421804Z [wss] HELO
2021-02-02T13:00:11.6427319Z GOODBYE
2021-02-02T13:00:11.6436840Z [wss] HELO
2021-02-02T13:00:11.6454191Z 
2021-02-02T13:00:11.6455897Z > [tls] Interact with web socket wss passthrough
2021-02-02T13:00:11.6497882Z spawn docker run --rm -ti --network=host solsson/websocat --insecure wss://localhost:30000
2021-02-02T13:00:12.6591804Z HELO
2021-02-02T13:00:12.6596911Z [wss] HELO
2021-02-02T13:00:12.6601501Z GOODBYE
2021-02-02T13:00:12.6603025Z [wss] HELO
2021-02-02T13:00:12.6694373Z 
2021-02-02T13:00:12.6695740Z > [tls] Cleanup (.)
```

While on mine and others, it does:
```
2021-02-04T11:03:38.1149607Z > [tls] Interact with web socket ws -> ws
2021-02-04T11:03:38.3143654Z spawn docker run --rm -ti --network=host solsson/websocat ws://localhost:10000
2021-02-04T11:03:38.5956163Z websocat: WebSocketError: I/O failure
2021-02-04T11:03:38.5957714Z websocat: error running
2021-02-04T11:03:38.5967854Z HELO
2021-02-04T11:03:38.6510123Z send: spawn id exp3 not open
2021-02-04T11:03:38.6511546Z     while executing
2021-02-04T11:03:38.6512455Z "send \x03"
2021-02-04T11:03:38.6542212Z 
2021-02-04T11:03:38.6543324Z > [tls] Cleanup (.)
```

This seems to be an IO error unrelated to the current work. What can I do about it?

**Comment by asraa** [comment]
Hmm thanks @plallin for re-running! I think verify-examples has been flakey, and those are most likely unrelated to your code since it's just the log message changed.

> I've not been able to execute them locally yet

For running locally (assuming there's enough space to build), check out https://github.com/envoyproxy/envoy/blob/17e815122ff53d0ac6cb2d64cdbf1bfc547bb7e8/bazel/README.md#testing-envoy-with-bazel once you've gone through the quick start.

I'll keep an eye on verify-examples.

**Review by asraa — APPROVED** [review]

**Comment by plallin** [comment]
@asraa Thanks! I've been able to run the test locally since; unfortunately my personal computer isn't optimised for development and I run into an OOM. I can probably configure Bazel to not use as much memory, but that's a problem for a future PR :-)

I have checked other PRs here, they run in the same issue with verify-examples, see [here]( https://github.com/envoyproxy/envoy/pull/14907) for example

**Comment by asraa** [comment]
Thanks! @envoyproxy/senior-maintainers can you  please merge this PR? It changes log messages in the H/1 codec

**Pull request merged**